### PR TITLE
Adapt GetHashRange to AllNodesRequestCore and new request framework

### LINKED ISSUE #70: Convert client GetHashRange handler to use AllNodesRequestCore
Also adapt to use the new supported codes from swarm.

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
Fixes #70.

### CODE REVIEW

File: src/dhtproto/node/neo/request/GetHashRange.d [hunk 1]
```diff
@@ -116,15 +116,8 @@ public abstract scope class GetHashRangeProtocol_v0
         scope ( exit )
             this.unregisterForHashRangeUpdates();
 
-        // Send Started status code and current hash range
+        // Send current hash range
```
Comment (gavin-norman-sociomantic) [inline src/dhtproto/node/neo/request/GetHashRange.d:121]: Note: sending the status code is removed as the error code was never sent.

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 2]
```diff
@@ -208,229 +193,197 @@ public struct GetHashRange
...
 
-    public this ( RequestOnConn.EventDispatcherAllNodes conn,
-                  void[] context_blob )
+    private void disconnected ( Exception e )
     {
-        this.conn = conn;
-        this.context = GetHashRange.getContext(context_blob);
+        // Log errors, except when the connection was deliberately shut down.
+        if ( cast(Connection.ConnectionClosedException)e is null )
+            log.error("I/O error: {} @ {}:{}", getMsg(e), e.file, e.line);
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d]: `e.message()` will now be fine :-).

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 3]
```diff
@@ -208,229 +193,197 @@ public struct GetHashRange
...
-            }
-        );
-        this.conn.flush();
-
-        // Receive status from node and stop the request if not Ok
-        auto status = conn.receiveValue!(StatusCode)();
-        switch ( status )
+        auto supported = cast(SupportedStatus)code;
+        if ( !GetHashRange.handleSupportedCodes(supported,
+            this.context, this.conn.remote_address) )
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d]: Perhaps in the next swarm we can adapt the `AllNodesRequest` to pass the `SupportedStatus` to this policy method.

Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: (or adapt it to avoid the policy all together).

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: Yeah, we can probably provide something standard.

**Comment by gavin-norman-sociomantic** [comment]
(Fixing D2 errors at present.)

**Comment by gavin-norman-sociomantic** [comment]
Updated.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Rebased.

**Comment by gavin-norman-sociomantic** [comment]
Should be rebased on https://github.com/sociomantic-tsunami/dhtproto/pull/94 and adapted to use the same framework in swarm.

**Comment by gavin-norman-sociomantic** [comment]
Rebased on #94. Still needs to be updated to use the new node request handler framework.

**Comment by gavin-norman-sociomantic** [comment]
>Still needs to be updated to use the new node request handler framework.

Done.

**Comment by gavin-norman-sociomantic** [comment]
#94 merged. Unblocked.

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 4]
```diff
@@ -208,229 +193,199 @@ public struct GetHashRange
...
-                payload.add(GetHashRange.cmd.ver);
-            }
-        );
-        this.conn.flush();
-
-        // Receive status from node and stop the request if not Ok
-        auto status = conn.receiveValue!(StatusCode)();
-        switch ( status )
+        auto supported = cast(SupportedStatus)code;
+        if ( !GetHashRange.handleSupportedCodes(supported,
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d]: This diff makes it really hard to spot, but:

```
return GetHashRange.handleSupportedCodes(supported,
    this.context, this.conn.remote_address);
```

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: Indeed, well spotted. Updated.

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 5]
```diff
@@ -208,229 +193,194 @@ public struct GetHashRange
...
+                            this.conn.message_parser.parseBody(payload_slice,
+                                new_min, new_max);
+
+                            this.resources.node_hash_ranges.updateNodeHashRange(
+                                this.conn.remote_address, new_min, new_max);
+                            break;
+
+                        default:
+                            log.error("Unknown message code {} received", msg_type);
+                            this.conn.shutdownWithProtocolError("Message parsing error");
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d]: Do you need to throw here?

```
throw this.conn.shutdownWithProtocolError(...)
```

?

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: OHHH! Shit. I didn't know that method _returns_ an exception, rather than actually throwing one :grimacing:

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: See https://github.com/sociomantic-tsunami/swarm/pull/292.

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d]: Updated.

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 6]
```diff
@@ -208,229 +193,195 @@ public struct GetHashRange
...
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            this.conn.remote_address, new_min, new_max);
-                        break;
-
-                    default:
-                        log.error("Unknown message code {} received", msg_type);
-                        this.conn.shutdownWithProtocolError("Message parsing error");
-                }
+                Const!(void)[] payload_slice = payload;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d:339]: Out of curiosity, why is this needed? Is the `payload` not a lvalue (since `getValue` will modify it?). Would it work if you made `(Const!(void)[] payload)`?

File: src/dhtproto/client/request/internal/GetHashRange.d [hunk 7]
```diff
@@ -208,229 +193,195 @@ public struct GetHashRange
...
+        this.resources.node_hash_ranges.updateNodeHashRange(
+            this.conn.remote_address, min, max);
 
-        return State.Receiving;
+        while ( true )
+        {
+            this.conn.receive(
+                ( in void[] payload )
+                {
+                    Const!(void)[] payload_slice = payload;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/GetHashRange.d:352]: ... why I asked the previous question is because this is (as expected) very common thing to do - to use `MessageParser` inside receive delegate.

**Review by nemanja-boric-sociomantic — APPROVED** [review]
My questions were not related to this PR, but were more general. LGTM.

File: src/dhtproto/client/request/internal/GetHashRange.d (hunk 6 above)
Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d:339]: It's because `MessageParser.getValue` modifies the slice that is passed to it. It's not possible to mutate the `in void[] payload`.

Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d:339]: That's right, but what about if the delegate would accept `Const!(void)[] payload`? I see no reason to accept it as `in`.

Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d:339]: Oh I see what you mean. Yeah, modifying the content of the payload should be disallowed, but modifying the slice to the payload buffer is fine. Changing this delegate type would be a breaking change in swarm, though :(

Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/GetHashRange.d:339]: I love breaking changes ;-). All right, I'll create issue to target v6.0.0 in swarm.
